### PR TITLE
Remove Global Error handler.

### LINKED ISSUE #2175: Remove global error handler in favor of internal logs
- Internal logs support is now available (#2128), and it would make sense to use it for internal errors.
- This allows the removal of the global error handler.
- Current downside: it won’t be possible to register a custom handler for errors at this time. However, plan is to add support for registering custom handlers for internal logs.
- If "internal-logs" is enabled, tracing `error` macro would be used for handling error, else the `eprintln` would be used.

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #2175

Needs #2259 #2257 #2256  to be merged first.

## Changes

- Removed `global::set_error_handler` and `global::handle_error`, as the `global_handle_error` usage inside the opentelemetry crates has already been replaced with `global::otel_info`, `otel_warn`, `otel_debug` and `otel_error`

## Migration Guide:

> 📌Note: This setup requires the internal-logs feature to be enabled for the OpenTelemetry crates. This feature is enabled by default, so application should ensure not to disable it explicitly for these crates in its `Cargo.toml`.

This guide explains how to migrate from `global::set_error_handler` to a new setup with `otel_*` macros and Tokio tracing. If your application previously used `global::set_error_handler`, remove it as part of this migration. If not, this configuration ensures errors previously going to `eprintln!` are now handled through structured tracing.

**Step 1: Remove set_error_handler (if used)**
If your code currently sets a custom error handler using global::set_error_handler, remove this line and any associated custom handler function.
```rust
// Remove any custom error handler
// global::set_error_handler(custom_error_handler).unwrap();
```

**Step 2: Internal OpenTelemetry Logs Filter and Layer:**
Set up a layer to handle OpenTelemetry internal logs and stop them from propagating to tracing bridge/appender.

1. ***Internal OpenTelemetry Logs Filter and Layer:***
    - This layer captures logs with the opentelemetry prefix and prints them to eprintln!.
 
```rust
use tracing_subscriber::filter::filter_fn;
use tracing_subscriber::fmt::Layer;

let opentelemetry_layer = Layer::new()
    .with_writer(std::io::stderr)
    .with_filter(filter_fn(|metadata| metadata.target().starts_with("opentelemetry")));
```

2. ***Bridge Layer for OpenTelemetry Tracing (Excludes OpenTelemetry Internal Logs)::*** 
    - To prevent logs with the opentelemetry prefix from propagating further, apply a filter_fn in this layer that excludes any logs with the opentelemetry target prefix.
 ```rust
use opentelemetry_appender_tracing::layer;

let non_opentelemetry_filter = filter_fn(|metadata| !metadata.target().starts_with("opentelemetry"));
let otel_bridge_layer = layer::OpenTelemetryTracingBridge::new(&cloned_provider)
    .with_filter(non_opentelemetry_filter);
```

**Step 3: Register Layers in the Tracing Registry**
```rust
use tracing_subscriber::prelude::*;

tracing_subscriber::registry()
    .with(opentelemetry_layer)       // OpenTelemetry internal logs (filtered out)
    .with(otel_bridge_layer)         // OpenTelemetry tracing bridge
    .init();
```

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: opentelemetry/src/global/internal_logging.rs [hunk 1]
```diff
@@ -156,3 +166,39 @@ macro_rules! otel_error {
...
+    #[error("{0}")]
+    /// Other types of failures not covered by the variants above.
+    Other(String),
+}
+
+impl<T> From<PoisonError<T>> for Error {
+    fn from(err: PoisonError<T>) -> Self {
+        Error::Other(err.to_string())
+    }
+}
```
Comment (lalitb) [inline opentelemetry/src/global/internal_logging.rs:204]: These (line 170 to 204 ) would be eventually moved to otel-sdk as part of #1042

File: opentelemetry/src/global/internal_logging.rs [hunk 2]
```diff
@@ -1,4 +1,14 @@
 #![allow(unused_macros)]
+
+#[cfg(feature = "logs")]
+use crate::logs::LogError;
+#[cfg(feature = "metrics")]
+use crate::metrics::MetricError;
+use crate::propagation::PropagationError;
+#[cfg(feature = "trace")]
+use crate::trace::TraceError;
+use std::sync::PoisonError;
```
Comment (lalitb) [inline opentelemetry/src/global/internal_logging.rs:10]: These (line 3 to 10 ) would be eventually moved to otel-sdk as part of #1042

**Review by cijothomas — CHANGES_REQUESTED** [review]
Given this is breaking change, we need to add it to changelog. and also show a migration guide for anyone who were relying on this previously.

**Comment by lalitb** [comment]
> Given this is breaking change, we need to add it to changelog. and also show a migration guide for anyone who were relying on this previously.

Done/

File: examples/self-diagnostics/src/main.rs [hunk 3]
```diff
@@ -77,18 +61,34 @@ fn init_logger_provider() -> opentelemetry_sdk::logs::LoggerProvider {
...
+    // only if they haven't been logged before.
     let opentelemetry_filter = tracing_subscriber::filter::filter_fn(|metadata| {
-        metadata.target().starts_with("opentelemetry")
+        if metadata.target().starts_with("opentelemetry_") {
+            // Ignore debug level logs
+            if metadata.level() == &tracing::Level::DEBUG {
+                return false;
+            }
+            let err = OtelError::Other(metadata.target().to_string()); // Convert target to an OtelError variant
+            if !GLOBAL_ERROR_STATE.mark_as_seen(&err) {
```
Comment (cijothomas) [inline examples/self-diagnostics/src/main.rs:73]: I am not sure what is the purpose of this. I believe we were planning to remove such logic from self-diag example?

Comment (lalitb) [reply examples/self-diagnostics/src/main.rs]: Yes, have removed the logic now.

File: opentelemetry/CHANGELOG.md [hunk 4]
```diff
@@ -33,6 +33,12 @@ let counter = meter.u64_counter("my_counter").build();
   - Replaced `global::meter_with_version` with `global::meter_with_scope`
   - Added `global::tracer_with_scope`
 
+- **Breaking change**: [#2260](https://github.com/open-telemetry/opentelemetry-rust/pull/2260)
+  - Removed `global::set_error_handler` and `global::handle_error`. 
+  - `global::handle_error` usage inside the opentelemetry crates has been replaced with `global::otel_info`, `otel_warn`, `otel_debug` and `otel_error` macros based on the severity of the internal logs.
+  - The default behavior of `global::handle_error` was to log the error using `eprintln!`. With otel macro, user now need to configure the tokio tracing layer to capture these logs.
```
Comment (cijothomas) [inline opentelemetry/CHANGELOG.md:39]:
```suggestion
  - The default behavior of `global::handle_error` was to log the error using `eprintln!`. With otel macro, the internal logs get emitted via `tracing` macros of matching severity. Users now need to configure the `tracing` layer to capture these logs.
```

File: examples/self-diagnostics/src/main.rs [hunk 5]
```diff
@@ -64,46 +26,46 @@ fn init_logger_provider() -> opentelemetry_sdk::logs::LoggerProvider {
...
+        let target = metadata.target();
+
+        // Only allow ERROR logs from specific targets
+        (target.starts_with("hyper")
+            || target.starts_with("hyper_util")
+            || target.starts_with("hyper")
+            || target.starts_with("tonic")
+            || target.starts_with("tower")
+            || target.starts_with("reqwest")
+            || target.starts_with("opentelemetry_"))
```
Comment (cijothomas) [inline examples/self-diagnostics/src/main.rs:42]:
```suggestion
            || target.starts_with("opentelemetry"))
```

File: opentelemetry/CHANGELOG.md (hunk 4 above)
Comment (lalitb) [reply opentelemetry/CHANGELOG.md:39]: Nice :)

**Review by cijothomas — APPROVED** [review]
LGTM.
I'll try our self-diagnostics with few scenarios later and make suggestions, if any. (I am a customer of the internal logs !!)

**Comment by mdecimus** [comment]
Thanks, I've just opened #2992 where I also explained another minor issue related to `SdkLogRecord`.

**Pull request merged**